CONTRIBUTING.md

```diff
@@ -0,0 +1,316 @@
+`require`s should always be put at the top of the source file,
+just after the copyright header, and before module globals and constants,
+and grouped in the following order:
+
+1. Standard libraries
+2. Related third-party libraries
+3. Local libraries
+
+For example,
+
+```lua
+local type = type
+local table = { concat = table.concat, insert = table.insert }
+
+local awful = require"awful"
+
+local helpers = require"vicious.helpers"
+```
+
+### String Quotes
+
+In Lua, single-quoted strings and double-quoted strings are the same,
+so the choice is totally up to you, but please be consistent within a module.
+When a string contains single or double quote characters, however,
+use the other one to avoid backslashes in the string. It improves readability:
+
+```lua
+'"key": "value"'        -- good
+"\"key\": \"value\""    -- no good
+```
+
+It is preferable to add a newline immediately after the opening long bracket:
+
+```lua
+foo = [[
+this is a really,
+really,
+really long text]]
+```
+
+### Naming Conventions
+
+Avoid using the characters `l` (lowercase letter el), `O` (uppercase letter oh),
+or `I` (uppercase letter eye) as single character variable names.
+In some fonts, these characters are indistinguishable from the 1's and 0's.
+
+#### Constants
+
+Constants are usually defined on a module level
+and written in all capital letters with underscores separating words.
+Examples include `MAX_OVERFLOW` and `TOTAL`.
+
+
+#### Function and Variable Names
+
+Function names should be lowercase, with words separated by underscores
+as necessary to improve readability.
+
+Variable names follow the same convention as function names.
+
+When you find it difficult to give descriptive names,
+use the functions and variable anonymously.
+
+### Performance Tips
+
+Vicious is meant to be run as part of the Awesome window manager,
+thus any little overhead may defect the responsiveness of the UI.
+While Lua is famous for its performance, there are a few things
+one can do to make use of all of its power.
+
+**Never** use global variables.  This includes the standard libraries,
+which, again, must be localized before use.  Remember, every widget type
+is to be called repeatedly every few seconds.
+
+Use closures when possible:
+
+* Define constants on the module level.
+* Avoid re-fetching the values that are not not meant to change.
+
+However, declare a variable only when you need it, to avoid declaring it
+without an initial value (and therefore you seldom forget to initialize it).
+Moreover, you shorten the scope of the variable, which increases readability. 
+
+### Copyright Header
+
+Vicious is released under the GNU GNU General Public License
+version 2 or later and each contributor holds the copyright
+on their contributing.  To make this collective control effective,
+each source file must include a notice of the following format
+denoting the name of all authors
+
+```lua
+-- <one line to give the program's name and a brief idea of what it does.>
```

Comment:
What format should this follow?

Reply:
For widget type I think something like *foo widget type* (*for system bar* if it's not system independent) or *widget type providing blah blah blah*.  I had been adding the filenames to my other projects but then I realized a few things:

* The header sample included in the license is meant for single-file programs.  That's why one should give *the program's name*.  For multi-file ones, we already have the *This file is part of foobar*.
* GNU's projects don't actually add filenames to the header, e.g. take a look at [gcc](https://github.com/gcc-mirror/gcc/blob/master/gcc/c/c-lang.c).
* The filename can be space consuming, especially when the description can be verbose for some of our widget types.  Also if it serves any meaning of getting where it at, it should be the relative path to the file within the project, which is even longer.